File: Emulator/src/main/core/game/world/map/Viewport.java
package core.game.world.map;

import core.game.node.entity.Entity;
import core.game.node.entity.npc.NPC;
import core.game.node.entity.player.Player;

import java.util.LinkedList;
import java.util.List;

/**
 * The type Viewport.
 */
public final class Viewport {

    /**
     * The constant CHUNK_SIZE.
     */
    public static final int CHUNK_SIZE = 5;

	private Region region;

	private RegionChunk[][] chunks = new RegionChunk[CHUNK_SIZE][CHUNK_SIZE];

	private RegionPlane currentPlane;

	private List<RegionPlane> viewingPlanes = new LinkedList<>();

    /**
     * Instantiates a new Viewport.
     */
    public Viewport() {

	}

    /**
     * Update viewport.
     *
     * @param entity the entity
     */
    public void updateViewport(Entity entity) {
		RegionChunk chunk = RegionManager.getRegionChunk(entity.getLocation());
		int center = chunks.length >> 1;
		if (chunks[center][center] == chunk) {
			return;
		}
		int offset = center * -8;
		Location l = chunk.getCurrentBase();
		for (int x = 0; x < chunks.length; x++) {
			for (int y = 0; y < chunks[x].length; y++) {
				chunks[x][y] = RegionManager.getRegionChunk(l.transform(offset + (8 * x), offset + (8 * y), 0));
			}
		}
	}

    /**
     * Remove.
     *
     * @param entity the entity
     */
    public void remove(Entity entity) {
		if (region == null) {
			return;
		}
		if (entity instanceof Player) {
			region.remove((Player) entity);
			Region region = null;
			for (RegionPlane r : viewingPlanes) {
				if (region != r.getRegion()) {
					region = r.getRegion();
					region.decrementViewAmount();
					region.checkInactive();
				}
			}
		} else {
			region.remove((NPC) entity);
		}
	}

    /**
     * Gets region.
     *
     * @return the region
     */
    public Region getRegion() {
		return region;
	}

    /**
     * Sets region.
     *
     * @param region the region
     */
    public void setRegion(Region region) {
		this.region = region;
	}

    /**
     * Get chunks region chunk [ ] [ ].
     *
     * @return the region chunk [ ] [ ]
     */
    public RegionChunk[][] getChunks() {
		return chunks;
	}

    /**
     * Sets chunks.
     *
     * @param chunks the chunks
     */
    public void setChunks(RegionChunk[][] chunks) {
		this.chunks = chunks;
	}

    /**
     * Gets viewing planes.
     *
     * @return the viewing planes
     */
    public List<RegionPlane> getViewingPlanes() {
		return viewingPlanes;
	}

    /**
     * Sets viewing planes.
     *
     * @param regions the regions
     */
    public void setViewingPlanes(List<RegionPlane> regions) {
		this.viewingPlanes = regions;
	}

    /**
     * Gets current plane.
     *
     * @return the current plane
     */
    public RegionPlane getCurrentPlane() {
		return currentPlane;
	}

    /**
     * Sets current plane.
     *
     * @param currentPlane the current plane
     */
    public void setCurrentPlane(RegionPlane currentPlane) {
		this.currentPlane = currentPlane;
	}

}

File: Emulator/src/main/core/game/world/map/Viewport.kt
package core.game.world.map

import core.game.node.entity.Entity
import core.game.node.entity.npc.NPC
import core.game.node.entity.player.Player
import core.game.world.map.RegionManager.getRegionChunk
import java.util.*

/**
 * Represents an entity's viewport.
 *
 * @author Emperor
 */
class Viewport {
    /**
     * The region.
     */
    var region: Region? = null

    /**
     * The region chunks.
     */
    var chunks: Array<Array<RegionChunk?>> = Array(CHUNK_SIZE) { arrayOfNulls(CHUNK_SIZE) }

    /**
     * The region plane.
     */
    var currentPlane: RegionPlane? = null

    /**
     * The region planes the entity can see.
     */
    var viewingPlanes: List<RegionPlane> = LinkedList()

    /**
     * Updates the entity's viewport.
     *
     * @param entity The entity.
     */
    fun updateViewport(entity: Entity) {
        val chunk = getRegionChunk(entity.location)
        val center = chunks.size shr 1
        if (chunks[center][center] === chunk) {
            return
        }
        val offset = center * -8
        val l = chunk.getCurrentBase()
        for (x in chunks.indices) {
            for (y in chunks[x].indices) {
                chunks[x][y] = getRegionChunk(l.transform(offset + (8 * x), offset + (8 * y), 0))
            }
        }
    }

    /**
     * Removes the entity from the viewingPlanes.
     *
     * @param entity The entity.
     */
    fun remove(entity: Entity?) {
        if (region == null) {
            return
        }
        if (entity is Player) {
            region!!.remove(entity)
            var region: Region? = null
            for (r in viewingPlanes) {
                if (region !== r.region) {
                    region = r.region
                    region.decrementViewAmount()
                    region.checkInactive()
                }
            }
        } else {
            region!!.remove(entity as NPC?)
        }
    }

    companion object {
        /**
         * The amount of chunks in a viewport.
         */
        const val CHUNK_SIZE: Int = 5
    }
}
